<div style="background: black; padding: 10px 250px"><img src="https://www.veldikompetens.se/wp-content/themes/consid/static/icons/VeldiKompetens_Logo_Web_Negative.svg" title="Veldi kompetens" /></div>

<hr><h1><center>Assignment 3 - Working on a fresh dataset </center></h1>

<h3>Instructions </h3>
<p>In this assignment you will be examining a new dataset on different wine types. This is going to be a somewhat open-ended assignment where it is fairly easy to achieve a pass, but if you really want to dig into it and fully understand the data some hints will be given for that purpose as well. There will be a section marked with mandatory and one with optional. For handin and to pass you need to complete the mandatory section. </p>
<p>The dataset consists of characteristics of the wine and its assessed rating by professionals in the quality column. We will perform some simpler analysis operations on this dataset

Link for more info can be found here: https://www.kaggle.com/uciml/red-wine-quality-cortez-et-al-2009</p>
<h3> 1. Setup </h3>
<p>Neccesary libraries: </p>

In [80]:
import numpy as np
import pandas as pd

# Remeber you need to download the csv and place it into your notebook folder
df = pd.read_csv("winequality-red.csv")
df.shape


(1600, 12)

<h3> 2. The assignment</h3>
<h4> 2.1 Introduction to the dataset </h4>
<p>Lets first get a good grasp of the data represented use the describe function to generate a general overview and lets set two good to have variables; the amount of samples (the samples variable) and the amount of columns/attributes (the columns variable  </p>

In [81]:
samples = df.shape[0]
columns = len(df.columns)#df.shape[1]#TODO

print(samples)
print(columns)
#TODO Use describe to get a summarized view of the dataset
df.describe()

1600
12


fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

<div class="output"><div class="output_area"><div class="run_this_cell"></div><div class="prompt"></div><div class="output_subarea output_text output_stream output_stdout" dir="auto"><pre>1600
12
</pre></div></div><div class="output_area"><div class="run_this_cell"></div><div class="prompt output_prompt"><bdi>Out[2]:</bdi></div><div class="output_subarea output_html rendered_html output_result" dir="auto"><div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>fixed acidity</th>
      <th>volatile acidity</th>
      <th>citric acid</th>
      <th>residual sugar</th>
      <th>chlorides</th>
      <th>free sulfur dioxide</th>
      <th>total sulfur dioxide</th>
      <th>density</th>
      <th>pH</th>
      <th>sulphates</th>
      <th>alcohol</th>
      <th>quality</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>8.319637</td>
      <td>0.527821</td>
      <td>0.270976</td>
      <td>2.538806</td>
      <td>0.087467</td>
      <td>15.874922</td>
      <td>46.467792</td>
      <td>0.996747</td>
      <td>3.311113</td>
      <td>0.658149</td>
      <td>10.422983</td>
      <td>5.636023</td>
    </tr>
    <tr>
      <th>std</th>
      <td>1.741096</td>
      <td>0.179060</td>
      <td>0.194801</td>
      <td>1.409928</td>
      <td>0.047065</td>
      <td>10.460157</td>
      <td>32.895324</td>
      <td>0.001887</td>
      <td>0.154386</td>
      <td>0.169507</td>
      <td>1.065668</td>
      <td>0.807569</td>
    </tr>
    <tr>
      <th>min</th>
      <td>4.600000</td>
      <td>0.120000</td>
      <td>0.000000</td>
      <td>0.900000</td>
      <td>0.012000</td>
      <td>1.000000</td>
      <td>6.000000</td>
      <td>0.990070</td>
      <td>2.740000</td>
      <td>0.330000</td>
      <td>8.400000</td>
      <td>3.000000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>7.100000</td>
      <td>0.390000</td>
      <td>0.090000</td>
      <td>1.900000</td>
      <td>0.070000</td>
      <td>7.000000</td>
      <td>22.000000</td>
      <td>0.995600</td>
      <td>3.210000</td>
      <td>0.550000</td>
      <td>9.500000</td>
      <td>5.000000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>7.900000</td>
      <td>0.520000</td>
      <td>0.260000</td>
      <td>2.200000</td>
      <td>0.079000</td>
      <td>14.000000</td>
      <td>38.000000</td>
      <td>0.996750</td>
      <td>3.310000</td>
      <td>0.620000</td>
      <td>10.200000</td>
      <td>6.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>9.200000</td>
      <td>0.640000</td>
      <td>0.420000</td>
      <td>2.600000</td>
      <td>0.090000</td>
      <td>21.000000</td>
      <td>62.000000</td>
      <td>0.997835</td>
      <td>3.400000</td>
      <td>0.730000</td>
      <td>11.100000</td>
      <td>6.000000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>15.900000</td>
      <td>1.580000</td>
      <td>1.000000</td>
      <td>15.500000</td>
      <td>0.611000</td>
      <td>72.000000</td>
      <td>289.000000</td>
      <td>1.003690</td>
      <td>4.010000</td>
      <td>2.000000</td>
      <td>14.900000</td>
      <td>8.000000</td>
    </tr>
  </tbody>
</table>
</div></div></div></div>

<h4>2.2 Finding and dropping NA values </h4>
<p>Lets now check if there are any NA values in any of our columns </p>

In [82]:
# For each column in the dataset perform the isna() operation and summarize the missing samples
    # After having calculated The amount of samples that have na values, divide it by total sample size and mulitply by 100
    # To get the percentage
def count_nan(df):
    for col in df.columns:
        col_name = col
        missing = df[col].isna().sum()
        percent = missing/samples*100
        print(f"The column {col_name} is missing {percent}% of data! ")
        
count_nan(df)


The column fixed acidity is missing 0.0625% of data! 
The column volatile acidity is missing 0.0625% of data! 
The column citric acid is missing 0.0625% of data! 
The column residual sugar is missing 0.0625% of data! 
The column chlorides is missing 0.0625% of data! 
The column free sulfur dioxide is missing 0.0625% of data! 
The column total sulfur dioxide is missing 0.0625% of data! 
The column density is missing 0.0625% of data! 
The column pH is missing 0.0625% of data! 
The column sulphates is missing 0.0625% of data! 
The column alcohol is missing 0.0625% of data! 
The column quality is missing 0.0625% of data! 


Expected output:
<pre>The column fixed acidity is missing 0.0625% of data! 
The column volatile acidity is missing 0.0625% of data! 
The column citric acid is missing 0.0625% of data! 
The column residual sugar is missing 0.0625% of data! 
The column chlorides is missing 0.0625% of data! 
The column free sulfur dioxide is missing 0.0625% of data! 
The column total sulfur dioxide is missing 0.0625% of data! 
The column density is missing 0.0625% of data! 
The column pH is missing 0.0625% of data! 
The column sulphates is missing 0.0625% of data! 
The column alcohol is missing 0.0625% of data! 
The column quality is missing 0.0625% of data! 
</pre>

Next up is a bit of a tricky part; We will now identify every sample that isna but instead of summarizing we will identify the samples such that we later on can remove them. Once again use isna() but this time say all() instead of sum()! You might also need to specify the axis

In [83]:
# 1. Extract all the samples that meet the criteria of isna()
nan_samples = df[df.isna().all(axis=1)]#TODO
nan_samples
# 2. Get the indexs for these rows
nan_indexes = nan_samples.index.values
nan_indexes
# 3. Drop them from the dataframe
df_clean = df.dropna()#TODO

# Finally, lets rerun the count_nan function to examine our revised dataset
count_nan(df_clean)
#df.shape

The column fixed acidity is missing 0.0% of data! 
The column volatile acidity is missing 0.0% of data! 
The column citric acid is missing 0.0% of data! 
The column residual sugar is missing 0.0% of data! 
The column chlorides is missing 0.0% of data! 
The column free sulfur dioxide is missing 0.0% of data! 
The column total sulfur dioxide is missing 0.0% of data! 
The column density is missing 0.0% of data! 
The column pH is missing 0.0% of data! 
The column sulphates is missing 0.0% of data! 
The column alcohol is missing 0.0% of data! 
The column quality is missing 0.0% of data! 


Expected output: 
<pre>The column fixed acidity is missing 0.0% of data! 
The column volatile acidity is missing 0.0% of data! 
The column citric acid is missing 0.0% of data! 
The column residual sugar is missing 0.0% of data! 
The column chlorides is missing 0.0% of data! 
The column free sulfur dioxide is missing 0.0% of data! 
The column total sulfur dioxide is missing 0.0% of data! 
The column density is missing 0.0% of data! 
The column pH is missing 0.0% of data! 
The column sulphates is missing 0.0% of data! 
The column alcohol is missing 0.0% of data! 
The column quality is missing 0.0% of data! 
</pre>

<h4>2.3 Filtering and reindexing</h4>
<p>Next up we will filter the dataset by quality and then reindex the dataset </p>

In [85]:
df_sort = df.sort_values(by="quality",ascending=True)#TODO
df_sort_reind=df_sort.reset_index(drop=True)
#f = #TODO
df_sort_reind

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.1             0.875         0.05             5.7      0.082   
1              10.4             0.440         0.42             1.5      0.145   
2               8.3             1.020         0.02             3.4      0.084   
3               6.8             0.815         0.00             1.2      0.267   
4              11.6             0.580         0.66             2.2      0.074   
...             ...               ...          ...             ...        ...   
1595            7.2             0.330         0.33             1.7      0.061   
1596           10.0             0.260         0.54             1.9      0.083   
1597            7.2             0.380         0.31             2.0      0.056   
1598            7.4             0.360         0.30             1.8      0.074   
1599            NaN               NaN          NaN             NaN        NaN   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                     3.0                  14.0  0.99808  3.40       0.52   
1                    34.0                  48.0  0.99832  3.38       0.86   
2                     6.0                  11.0  0.99892  3.48       0.49   
3                    16.0                  29.0  0.99471  3.32       0.51   
4                    10.0                  47.0  1.00080  3.25       0.57   
...                   ...                   ...      ...   ...        ...   
1595                  3.0                  13.0  0.99600  3.23       1.10   
1596                 42.0                  74.0  0.99451  2.98       0.63   
1597                 15.0                  29.0  0.99472  3.23       0.76   
1598                 17.0                  24.0  0.99419  3.24       0.70   
1599                  NaN                   NaN      NaN   NaN        NaN   

      alcohol  quality  
0        10.2      3.0  
1         9.9      3.0  
2        11.0      3.0  
3         9.8      3.0  
4         9.0      3.0  
...       ...      ...  
1595     10.0      8.0  
1596     11.8      8.0  
1597     11.3      8.0  
1598     11.4      8.0  
1599      NaN      NaN  

[1600 rows x 12 columns]

Expected output: 

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>fixed acidity</th>
      <th>volatile acidity</th>
      <th>citric acid</th>
      <th>residual sugar</th>
      <th>chlorides</th>
      <th>free sulfur dioxide</th>
      <th>total sulfur dioxide</th>
      <th>density</th>
      <th>pH</th>
      <th>sulphates</th>
      <th>alcohol</th>
      <th>quality</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>7.1</td>
      <td>0.875</td>
      <td>0.05</td>
      <td>5.7</td>
      <td>0.082</td>
      <td>3.0</td>
      <td>14.0</td>
      <td>0.99808</td>
      <td>3.40</td>
      <td>0.52</td>
      <td>10.2</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>10.4</td>
      <td>0.440</td>
      <td>0.42</td>
      <td>1.5</td>
      <td>0.145</td>
      <td>34.0</td>
      <td>48.0</td>
      <td>0.99832</td>
      <td>3.38</td>
      <td>0.86</td>
      <td>9.9</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>8.3</td>
      <td>1.020</td>
      <td>0.02</td>
      <td>3.4</td>
      <td>0.084</td>
      <td>6.0</td>
      <td>11.0</td>
      <td>0.99892</td>
      <td>3.48</td>
      <td>0.49</td>
      <td>11.0</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>6.8</td>
      <td>0.815</td>
      <td>0.00</td>
      <td>1.2</td>
      <td>0.267</td>
      <td>16.0</td>
      <td>29.0</td>
      <td>0.99471</td>
      <td>3.32</td>
      <td>0.51</td>
      <td>9.8</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>11.6</td>
      <td>0.580</td>
      <td>0.66</td>
      <td>2.2</td>
      <td>0.074</td>
      <td>10.0</td>
      <td>47.0</td>
      <td>1.00080</td>
      <td>3.25</td>
      <td>0.57</td>
      <td>9.0</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1594</th>
      <td>10.7</td>
      <td>0.350</td>
      <td>0.53</td>
      <td>2.6</td>
      <td>0.070</td>
      <td>5.0</td>
      <td>16.0</td>
      <td>0.99720</td>
      <td>3.15</td>
      <td>0.65</td>
      <td>11.0</td>
      <td>8.0</td>
    </tr>
    <tr>
      <th>1595</th>
      <td>7.2</td>
      <td>0.330</td>
      <td>0.33</td>
      <td>1.7</td>
      <td>0.061</td>
      <td>3.0</td>
      <td>13.0</td>
      <td>0.99600</td>
      <td>3.23</td>
      <td>1.10</td>
      <td>10.0</td>
      <td>8.0</td>
    </tr>
    <tr>
      <th>1596</th>
      <td>10.0</td>
      <td>0.260</td>
      <td>0.54</td>
      <td>1.9</td>
      <td>0.083</td>
      <td>42.0</td>
      <td>74.0</td>
      <td>0.99451</td>
      <td>2.98</td>
      <td>0.63</td>
      <td>11.8</td>
      <td>8.0</td>
    </tr>
    <tr>
      <th>1597</th>
      <td>7.2</td>
      <td>0.380</td>
      <td>0.31</td>
      <td>2.0</td>
      <td>0.056</td>
      <td>15.0</td>
      <td>29.0</td>
      <td>0.99472</td>
      <td>3.23</td>
      <td>0.76</td>
      <td>11.3</td>
      <td>8.0</td>
    </tr>
    <tr>
      <th>1598</th>
      <td>7.4</td>
      <td>0.360</td>
      <td>0.30</td>
      <td>1.8</td>
      <td>0.074</td>
      <td>17.0</td>
      <td>24.0</td>
      <td>0.99419</td>
      <td>3.24</td>
      <td>0.70</td>
      <td>11.4</td>
      <td>8.0</td>
    </tr>
  </tbody>
</table>

<h4>2.4 Transforming the data</h4>
<p>Lastly as part of the assignment we will be transforming and normalizing the data values. In Exercise 3c you saw how to perform this for one column, here we will recreate it but for all the columns that can benefit from it (which truth be told is all of them). The solutions is also a bit different so you cant fully copy paste. In short; the objective is to normalize each column in the dataframe. Hint; Check out the pandas docs for an idea of how to use apply at the df[column] part!</p>

In [86]:
#def normalize_values(col_value, min_val, max_val):
#    normalized_val = np.divide((col_value - min_val), (max_val - min_val)) # TODO
#    return normalized_val

#def main():
#    for column in df.columns:
#        col_value = df[column]
#        min_val = col_value.min()#TODO
#        max_val = col_value.min()#TODO
#        df[column] = df[column].apply(normalize_values,args=(min_val,max_val))#TODO

#main()
#df.describe()


from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()  # Initialize MinMaxScaler
df_copy =df.copy()
numerical_cols = df_copy.select_dtypes(include="number").columns.tolist()

for col in numerical_cols:
    df_copy[col] = scaler.fit_transform(df_copy[[col]])
df_copy.reset_index()

index  fixed acidity  volatile acidity  citric acid  residual sugar  \
0         0       0.247788          0.397260         0.00        0.068493   
1         1       0.283186          0.520548         0.00        0.116438   
2         2       0.283186          0.438356         0.04        0.095890   
3         3       0.584071          0.109589         0.56        0.068493   
4         4       0.247788          0.397260         0.00        0.068493   
...     ...            ...               ...          ...             ...   
1595   1595       0.141593          0.328767         0.08        0.075342   
1596   1596       0.115044          0.294521         0.10        0.089041   
1597   1597       0.150442          0.267123         0.13        0.095890   
1598   1598       0.115044          0.359589         0.12        0.075342   
1599   1599       0.123894          0.130137         0.47        0.184932   

      chlorides  free sulfur dioxide  total sulfur dioxide   density  \
0      0.106845             0.140845              0.098940  0.567548   
1      0.143573             0.338028              0.215548  0.494126   
2      0.133556             0.197183              0.169611  0.508811   
3      0.105175             0.225352              0.190813  0.582232   
4      0.106845             0.140845              0.098940  0.567548   
...         ...                  ...                   ...       ...   
1595   0.130217             0.436620              0.134276  0.354626   
1596   0.083472             0.535211              0.159011  0.370778   
1597   0.106845             0.394366              0.120141  0.416300   
1598   0.105175             0.436620              0.134276  0.396476   
1599   0.091820             0.239437              0.127208  0.397944   

            pH  sulphates   alcohol  quality  
0     0.606299   0.137725  0.153846      0.4  
1     0.362205   0.209581  0.215385      0.4  
2     0.409449   0.191617  0.215385      0.4  
3     0.330709   0.149701  0.215385      0.6  
4     0.606299   0.137725  0.153846      0.4  
...        ...        ...       ...      ...  
1595  0.559055   0.149701  0.323077      0.4  
1596  0.614173   0.257485  0.430769      0.6  
1597  0.535433   0.251497  0.400000      0.6  
1598  0.653543   0.227545  0.276923      0.4  
1599  0.511811   0.197605  0.400000      0.6  

[1600 rows x 13 columns]

Expected output
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>fixed acidity</th>
      <th>volatile acidity</th>
      <th>citric acid</th>
      <th>residual sugar</th>
      <th>chlorides</th>
      <th>free sulfur dioxide</th>
      <th>total sulfur dioxide</th>
      <th>density</th>
      <th>pH</th>
      <th>sulphates</th>
      <th>alcohol</th>
      <th>quality</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
      <td>1599.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>0.329171</td>
      <td>0.279329</td>
      <td>0.270976</td>
      <td>0.112247</td>
      <td>0.125988</td>
      <td>0.209506</td>
      <td>0.142996</td>
      <td>0.490211</td>
      <td>0.449695</td>
      <td>0.196496</td>
      <td>0.311228</td>
      <td>0.527205</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.154079</td>
      <td>0.122644</td>
      <td>0.194801</td>
      <td>0.096570</td>
      <td>0.078573</td>
      <td>0.147326</td>
      <td>0.116238</td>
      <td>0.138571</td>
      <td>0.121564</td>
      <td>0.101501</td>
      <td>0.163949</td>
      <td>0.161514</td>
    </tr>
    <tr>
      <th>min</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>0.221239</td>
      <td>0.184932</td>
      <td>0.090000</td>
      <td>0.068493</td>
      <td>0.096828</td>
      <td>0.084507</td>
      <td>0.056537</td>
      <td>0.406021</td>
      <td>0.370079</td>
      <td>0.131737</td>
      <td>0.169231</td>
      <td>0.400000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>0.292035</td>
      <td>0.273973</td>
      <td>0.260000</td>
      <td>0.089041</td>
      <td>0.111853</td>
      <td>0.183099</td>
      <td>0.113074</td>
      <td>0.490455</td>
      <td>0.448819</td>
      <td>0.173653</td>
      <td>0.276923</td>
      <td>0.600000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>0.407080</td>
      <td>0.356164</td>
      <td>0.420000</td>
      <td>0.116438</td>
      <td>0.130217</td>
      <td>0.281690</td>
      <td>0.197880</td>
      <td>0.570117</td>
      <td>0.519685</td>
      <td>0.239521</td>
      <td>0.415385</td>
      <td>0.600000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
    </tr>
  </tbody>
</table>

<h4>3. Closing Remarks</h4>
<p>Good job! You are now done with this weeks assignment. There is a lot more you can do with the data but these are the neccesary basics. In future courses we may very well introduce some more machine learning associated koncepts like kNN which is a method for estimating predications, for example the algorithm would learn from the represented numbers here a pattern to guess the quality of a wine from its attributes. However, that will not be part of this course. In the next week we will be simulating a pandemic, modelling eco systems and generating a dataset of our own!</p>


<h4> 4. If you want to do more </h4>
<p>These are just some suggestions if you want to learn more about working with data and are not mandatory for the assignment but can be fun to do: </p>

<ul>
    <li>Try out some plotting techniques (google pandas.boxplot for example)</li>
    <li>Add outliers manually to the dataset and try to find them via either plots, statistics or other means of analysis (google the concept of outliers! There is a ton of interesting facts to get through to help you find outliers) </li>
    </ul>

In [77]:
import matplotlib.pyplot as plt

In [90]:
df[["pH","alcohol"]].boxplot()
outlier = {"pH":[8],"alcohol":[4]}
df=df.append(pd.DataFrame(outlier),ignore_index=True)

C:\Users\xli204\AppData\Local\Temp\ipykernel_27032\4258031495.py:3: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df=df.append(pd.DataFrame(outlier),ignore_index=True)
